_posts/2024-06-30-concurrent-search-follow-up.md

```diff
@@ -0,0 +1,307 @@
+| `range-date-histo`                            | 1803                | 948                                        | 47%           | 1029                                    | 43%           |
+| `range-numeric-significant-terms`             | 2597                | 2951                                       | -14%          | 1627                                    | 37%           |
+
+#### System metrics
+
+|               Metric              | CS disabled | CS enabled (Lucene default slices) | CS enabled (fixed slice count=4) |
+|-----------------------------------|-------------|------------------------------------|----------------------------------|
+| p90 CPU                           | 25%         | 99%                                | 50%                              |
+| p90 JVM                           | 60%         | 60%                                | 60%                              |
+| Max `index_searcher` active threads | --          | 16                                 | 4                                |
+
+The performance results for these aggregation types confirm the results we saw with previous cluster configurations. We see strong performance improvements in most aggregation types, and again, this performance improvement diminishes and sometimes even regresses as we increase  concurrency under the CPU load.
+
+## Observations
+
+Increases or decreases in performance related to concurrent segment search can usually be attributed to one of the following four reasons.
+
+First, whenever the number of segment slices is large, the `index_searcher` thread pool fills. Whenever there are no threads available to execute the shard search task for a slice, the slice waits in the queue until other slices are finished processing. For example, in **Cluster setup 3** there are `4 * 17 = 68` total segment slices when using the Lucene default slice count but only 64 threads available in the concurrent search thread pool. Thus, 4 segment slices will spend some time waiting in the thread pool queue.
+
+Second, whenever the number of active threads is higher than the number of CPU cores, each individual thread may spend more time processing because the CPU cores are multiplexing tasks. By default, the `r5.2xlarge` instance with 8 CPU cores has 16 threads in the `index_searcher` thread pool and 13 threads in the search thread pool. If all 29 threads are concurrently processing search tasks, then each individual thread will encounter a longer processing time because there are only 8 CPU cores to serve these 29 threads.
+
+Third, the specific query implementation can greatly impact the performance when increasing concurrency because some queries may perform more duplicate work as the number of slices increases. For example, significant terms aggregations run count queries for each bucket key to determine the term background frequencies. Thus, duplicated bucket keys across segment slices result in duplicated count queries across slices as well.
```

Comment:
```suggestion
- Third, the specific query implementation can greatly impact performance when increasing concurrency because some queries may perform more duplicate work as the number of slices increases. For example, significant terms aggregations run count queries for each bucket key to determine the term background frequencies. Thus, duplicated bucket keys across segment slices result in duplicated count queries across slices as well.
```